### PR TITLE
[Classrooms] building numbers, link to facilities

### LINKED ISSUE #6592: link classrooms to facilities building pages
# Story
As a Classrooms site owner, I would like links to buildings to go to the building pages on Facilities rather than having building pages on the Classrooms website.

# Additional Info

It was decided that Facilities is going to add hours info from Sims into Bizhub which will allow us to output hours on the Facilities website. Since Facilities will have the hours listed at some point, Classrooms is ok with just linking to those building pages.

We need to get the building number since we only have the building abbreviation and building name currently. The building number doesn't currently exist in the endpoint we are using to get the list of buildings.

Current endpoint to get buildings https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraBldgRmCompleteList/list

**Possible other endpoints**

- Could maybe get building number with an additional call to https://bizhub.facilities.uiowa.edu/bizhub/ext/buildings
- Could maybe change endpoints and get list of buildings from https://api.maui.uiowa.edu/maui/api/pub/facilityBuildingRoom/list or
https://api.maui.uiowa.edu/maui/api/pub/facilityBuildingRoom/list?roomCategory=UNIVERSITY_CLASSROOM and https://api.maui.uiowa.edu/maui/api/pub/facilityBuildingRoom/list?roomCategory=PROGRAMMED_CLASSROOM
    - Need to understand if this endpoint contains same rooms as https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraBldgRmCompleteList/list plus our logic to filter out some categories

# Proposed Solution

- Get building number and store on config entities
- Update links on rooms to go to facilities building pages

### PR DESCRIPTION
**Pull request by cory-skeers**
Resolves #6592 

# How to test

```
git checkout classrooms-building-numbers && git pull && ddev blt ds --site=classrooms.uiowa.edu && ddev drush @classrooms.local uli /admin/structure/building
```
Check for a clean sync
Spot check some buildings and see that they have building numbers
Check that the following buildings aren't available for selection in creating rooms, filtering searches, etc.
- Calvin Hall
- South Quad
- Pharmaceutical Sciences Research Building
- Kirkwood Regional Center

```
ddev drush @classrooms.local classrooms-buildings
```
No new buildings should be imported (assuming the api results haven't changed by testing time)

Delete a building or multiple buildings
```
ddev drush @classrooms.local classrooms-buildings
```
The message should indicate that the same number of buildings you deleted should now be created 
Check the building(s) that you'd deleted, and see that they have been created, including their building number

Check some Room nodes, and see that the Building (toward the bottom of the display) is linked, and that link points to the correct building page on facilities.uiowa.edu.

Check that the Room teaser card building links are also correct (such as on https://classrooms.uiowa.ddev.site/find-university-classroom)

This work moves from the previous endpoint https://api.maui.uiowa.edu/maui/api/pub/registrar/courses/AstraBldgRmCompleteList/list with a custom filter to using https://api.maui.uiowa.edu/maui/api/pub/facilityBuildingRoom/list?roomCategory=UNIVERSITY_CLASSROOM and https://api.maui.uiowa.edu/maui/api/pub/facilityBuildingRoom/list?roomCategory=PROGRAMMED_CLASSROOM.
The previous filters were on the room "RegionList" property in the full list, filtered down to only:
```
        '1) University Classrooms - Level 1',
        '1) University Classrooms - Original',
        '1) University Classrooms',
        '1) University Classrooms - Study Space',
        '1) Programmed Classrooms - Level 2',
        'Classroom-Programmed',
```
As far as I can tell, we're getting the same end results with this switch (comparison of resultant buildings of the building process, as well as spot checking rooms contained between the two lists), but any extra checking here, or other methods of comparison are welcome.

I thought of writing a script to do our old filtering and compare the results with a full pull of the other two APIs for a thorough comparison, but haven't gone that far yet. It could be an option, though.

## Other changes

The RoomSchedule class was updated to use dependency injection for its services.
Spot check some rooms and compare their schedules with the room on classrooms.prod.drupal.uiowa.edu to spot check for regressions. (Largely this should be a check that it still functions--the services affected are those to get the current node form the path, to load the maui api, and to create a pager, so it's pretty much a "this still works" or "this is all broken" type update)

### CODE REVIEW

**Comment by joewhitsitt** [comment]
>Spot check some buildings and see that they have building numbers (There are 4 which don't, including Calvin Hall, South Quad, Pharmaceutical Sciences Research Building, and Kirkwood Regional Center--that is to be expected)

@cory-skeers we just got out of a meeting with the Classrooms staff. These buildings were mentioned as recently changed. By not having a building number in this new endpoint, it might mean that their config can be deleted.

We told them going forward we would send a quick email their way about any buildings being removed and they would do the same so we can update it to keep the filter dropdown up to date.

@briand44 any other thoughts?

**Comment by cory-skeers** [comment]
> > Spot check some buildings and see that they have building numbers (There are 4 which don't, including Calvin Hall, South Quad, Pharmaceutical Sciences Research Building, and Kirkwood Regional Center--that is to be expected)
> 
> @cory-skeers we just got out of a meeting with the Classrooms staff. These buildings were mentioned as recently changed. By not having a building number in this new endpoint, it might mean that their config can be deleted.
> 
> We told them going forward we would send a quick email their way about any buildings being removed and they would do the same so we can update it to keep the filter dropdown up to date.
> 
> @briand44 any other thoughts?

I've been thinking about this and how there are still rooms on the site that use these buildings. Deleting the buildings without removing the rooms could lead to errors, since the room will still be referencing an entity that no longer exists. So I think we'd either want to manually remove those rooms, or wait for them to remove them prior to removing the config.

If the buildings weren't config entities, we could add to the building update process to also remove buildings that aren't in the API and that aren't referenced by any rooms, so the filter would be up-to-date as long as the rooms are. But I think when we originally discussed the pros/cons of using config entities, part of it was that we don't have buildings going up and coming down all that often.

So basically...I think removing the config for those four buildings sounds good to me; we just need to know if it's fine to get rid of any rooms that are currently referencing them (and/or have classrooms folks remove them and let us know when  they're cleaned up).

**Comment by GaryRidgway** [comment]
Testing this, I found that if you try and access a room for a building without a number, then the site whitescreens (eg. https://classrooms.uiowa.ddev.site/calh-208)

<details><summary>See the error output here:</summary>
```
The website encountered an unexpected error. Please try again later.
Error: Call to a member function get() on null in classrooms_core_node_view() (line 131 of sites/classrooms.uiowa.edu/modules/classrooms_core/classrooms_core.module).
classrooms_core_node_view(Array, Object, Object, 'full')
call_user_func_array(Object, Array) (Line: 426)
Drupal\Core\Extension\ModuleHandler->Drupal\Core\Extension\{closure}(Object, 'classrooms_core') (Line: 405)
Drupal\Core\Extension\ModuleHandler->invokeAllWith('node_view', Object) (Line: 433)
Drupal\Core\Extension\ModuleHandler->invokeAll('node_view', Array) (Line: 288)
Drupal\Core\Entity\EntityViewBuilder->buildMultiple(Array) (Line: 239)
Drupal\Core\Entity\EntityViewBuilder->build(Array)
call_user_func_array(Array, Array) (Line: 101)
Drupal\Core\Render\Renderer->doTrustedCallback(Array, Array, 'Render #pre_render callbacks must be methods of a class that implements \Drupal\Core\Security\TrustedCallbackInterface or be an anonymous function. The callback was %s. See https://www.drupal.org/node/2966725', 'exception', 'Drupal\Core\Render\Element\RenderCallbackInterface') (Line: 772)
Drupal\Core\Render\Renderer->doCallback('#pre_render', Array, Array) (Line: 363)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array, ) (Line: 241)
Drupal\Core\Render\MainContent\HtmlRenderer->Drupal\Core\Render\MainContent\{closure}() (Line: 564)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 242)
Drupal\Core\Render\MainContent\HtmlRenderer->prepare(Array, Object, Object) (Line: 132)
Drupal\Core\Render\MainContent\HtmlRenderer->renderResponse(Array, Object, Object) (Line: 90)
Drupal\Core\EventSubscriber\MainContentViewSubscriber->onViewRenderArray(Object, 'kernel.view', Object)
call_user_func(Array, Object, 'kernel.view', Object) (Line: 142)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch(Object, 'kernel.view') (Line: 174)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 81)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 58)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 51)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 709)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```
</details> 

additionally, when i hit https://classrooms.uiowa.ddev.site/find-university-classroom it whitescreened as well

<details><summary>See the error output here:</summary>
```
The website encountered an unexpected error. Please try again later.
Error: Call to a member function get() on null in Drupal\classrooms_core\Entity\ClassroomsRoom->buildCard() (line 71 of sites/classrooms.uiowa.edu/modules/classrooms_core/src/Entity/ClassroomsRoom.php).
Drupal\classrooms_core\Entity\ClassroomsRoom->buildCard(Array) (Line: 166)
uiowa_core_entity_view(Array, Object, Object, 'teaser')
call_user_func_array(Object, Array) (Line: 426)
Drupal\Core\Extension\ModuleHandler->Drupal\Core\Extension\{closure}(Object, 'uiowa_core') (Line: 405)
Drupal\Core\Extension\ModuleHandler->invokeAllWith('entity_view', Object) (Line: 433)
Drupal\Core\Extension\ModuleHandler->invokeAll('entity_view', Array) (Line: 289)
Drupal\Core\Entity\EntityViewBuilder->buildMultiple(Array) (Line: 239)
Drupal\Core\Entity\EntityViewBuilder->build(Array)
call_user_func_array(Array, Array) (Line: 101)
Drupal\Core\Render\Renderer->doTrustedCallback(Array, Array, 'Render #pre_render callbacks must be methods of a class that implements \Drupal\Core\Security\TrustedCallbackInterface or be an anonymous function. The callback was %s. See https://www.drupal.org/node/2966725', 'exception', 'Drupal\Core\Render\Element\RenderCallbackInterface') (Line: 772)
Drupal\Core\Render\Renderer->doCallback('#pre_render', Array, Array) (Line: 363)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array) (Line: 479)
Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1) (Line: 60)
__TwigTemplate_5858a757a0e30d2a36a34d8ffde0a023->doDisplay(Array, Array) (Line: 405)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 378)
Twig\Template->display(Array) (Line: 390)
Twig\Template->render(Array) (Line: 55)
twig_render_template('core/themes/classy/templates/views/views-view-unformatted.html.twig', Array) (Line: 384)
Drupal\Core\Theme\ThemeManager->render('views_view_unformatted', Array) (Line: 422)
Drupal\Core\Render\Renderer->doRender(Array) (Line: 435)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array) (Line: 479)
Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1) (Line: 147)
__TwigTemplate_3a4967275c9227507a084050061e4e83->doDisplay(Array, Array) (Line: 405)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 378)
Twig\Template->display(Array) (Line: 390)
Twig\Template->render(Array) (Line: 55)
twig_render_template('themes/custom/uids_base/templates/views/views-view.html.twig', Array) (Line: 384)
Drupal\Core\Theme\ThemeManager->render('views_view', Array) (Line: 422)
Drupal\Core\Render\Renderer->doRender(Array) (Line: 435)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array) (Line: 581)
Drupal\Core\Template\TwigExtension->renderVar(Array) (Line: 44)
__TwigTemplate_d0ef55e18987bf3e608bf2b2d2a34014->doDisplay(Array, Array) (Line: 405)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 378)
Twig\Template->display(Array) (Line: 390)
Twig\Template->render(Array) (Line: 55)
twig_render_template('themes/custom/uids_base/templates/block/block.html.twig', Array) (Line: 384)
Drupal\Core\Theme\ThemeManager->render('block', Array) (Line: 422)
Drupal\Core\Render\Renderer->doRender(Array) (Line: 435)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array) (Line: 581)
Drupal\Core\Template\TwigExtension->renderVar(Array) (Line: 60)
__TwigTemplate_a8541e223cd3ad0d17f84e4e9fba806e->doDisplay(Array, Array) (Line: 405)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 378)
Twig\Template->display(Array) (Line: 50)
__TwigTemplate_a1d4f07a2257250568a52f3f4dde049c->doDisplay(Array, Array) (Line: 405)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 378)
Twig\Template->display(Array) (Line: 390)
Twig\Template->render(Array) (Line: 55)
twig_render_template('themes/custom/uids_base/templates/layouts/layout--onecol.html.twig', Array) (Line: 384)
Drupal\Core\Theme\ThemeManager->render('layout__onecol', Array) (Line: 422)
Drupal\Core\Render\Renderer->doRender(Array) (Line: 435)
Drupal\Core\Render\Renderer->doRender(Array) (Line: 435)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array) (Line: 479)
Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1) (Line: 115)
__TwigTemplate_c8aaa7b75f5fdd13b041744e7686d1a9->doDisplay(Array, Array) (Line: 405)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 378)
Twig\Template->display(Array) (Line: 390)
Twig\Template->render(Array) (Line: 55)
twig_render_template('themes/custom/uids_base/templates/content/node--page--full.html.twig', Array) (Line: 384)
Drupal\Core\Theme\ThemeManager->render('node', Array) (Line: 422)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 201)
Drupal\Core\Render\Renderer->render(Array, ) (Line: 241)
Drupal\Core\Render\MainContent\HtmlRenderer->Drupal\Core\Render\MainContent\{closure}() (Line: 564)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 242)
Drupal\Core\Render\MainContent\HtmlRenderer->prepare(Array, Object, Object) (Line: 132)
Drupal\Core\Render\MainContent\HtmlRenderer->renderResponse(Array, Object, Object) (Line: 90)
Drupal\Core\EventSubscriber\MainContentViewSubscriber->onViewRenderArray(Object, 'kernel.view', Object)
call_user_func(Array, Object, 'kernel.view', Object) (Line: 142)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch(Object, 'kernel.view') (Line: 174)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 81)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 58)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 51)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 709)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```
</details> 


For checking the classroom schedules, I was having a hard time finding a room with scheduled events. Are there any examples that you know of off the top of your head @cory-skeers ?

**Comment by cory-skeers** [comment]
> Testing this, I found that if you try and access a room for a building without a number, then the site whitescreens (eg. https://classrooms.uiowa.ddev.site/calh-208)

Shoot. I was hoping the no numbers buildings wouldn't get in the way of testing ahead of time.

We're likely going to be removing those buildings, but the rooms will need to go first. We'll be reaching out to them to confirm (https://github.com/uiowa/uiowa/issues/6592#issuecomment-1631052677) but as long as we can move forward with that, then we should be good there.

> additionally, when i hit https://classrooms.uiowa.ddev.site/find-university-classroom it whitescreened as well

Similar to the above. But for both, this is probably an indicator that there could be some better defensive coding put into place... I can take a look at that as well.

> For checking the classroom schedules, I was having a hard time finding a room with scheduled events. Are there any examples that you know of off the top of your head @cory-skeers ?

https://classrooms.prod.drupal.uiowa.edu/abw-240 and https://classrooms.prod.drupal.uiowa.edu/abw-116 each have at least one event. I usually have good luck with the ABW auditoriums.

**Comment by cory-skeers** [comment]
@GaryRidgway I added two checks for when the building is no longer there that should prevent the WSOD.

An email has also gone out to confirm that those buildings can be fully removed which'll help, but I think it'll be good to have the quick checks in place just in case things are out of alignment again in the future.

**Comment by GaryRidgway** [comment]
Wonderful! That sounds like it will help a lot. I will un-assign myself for now until I can be sure that I have time to test, but I am happy to see the progress!

**Comment by GaryRidgway** [comment]
After Cory's fixes, this looks good to go!

**Review by GaryRidgway — APPROVED** [review]
